contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,639 @@
+        shares = queue[lastId].cumulativeShares;
+
+        if (firstId > 0) {
+            eth -= queue[firstId - 1].cumulativeEther;
+            shares -= queue[firstId - 1].cumulativeShares;
+        }
+
+        eth = _min(eth, (shares * _totalPooledEther) / _totalShares);
+    }
+
+    function _isPriceHintValid(uint256 _requestId, uint256 hint) internal view returns (bool isInRange) {
+        uint256 hintLastId = finalizationPrices[hint].index;
+
+        isInRange = _requestId <= hintLastId;
+        if (hint > 0) {
+            uint256 previousId = finalizationPrices[hint - 1].index;
+
+            isInRange = isInRange && previousId < _requestId;
+        }
+    }
+
+    function _updatePriceHistory(uint128 _totalPooledEther, uint128 _totalShares, uint256 index) internal {
+        if (finalizationPrices.length == 0) {
+            finalizationPrices.push(Price(_totalPooledEther, _totalShares, index));
+        } else {
+            Price storage lastPrice = finalizationPrices[finalizationPrices.length - 1];
+
+            if (_totalPooledEther / _totalShares == lastPrice.totalPooledEther / lastPrice.totalShares) {
+                lastPrice.index = index;
+            } else {
+                finalizationPrices.push(Price(_totalPooledEther, _totalShares, index));
+            }
+        }
+    }
+
+    function _min(uint128 a, uint128 b) internal pure returns (uint128) {
+        return a < b ? a : b;
+    }
+
+    function _sendValue(address payable recipient, uint256 amount) internal {
+        if (address(this).balance < amount) revert NotEnoughEther();
+
+        // solhint-disable-next-line
+        (bool success, ) = recipient.call{value: amount}("");
+        if (!success) revert CantSendValueRecipientMayHaveReverted();
+    }
+
+    function _toUint64(uint256 value) internal pure returns (uint64) {
+        if (value > type(uint64).max) revert SafeCastValueDoesNotFit96Bits();
+        return uint64(value);
+    }
+
+    function _toUint128(uint256 value) internal pure returns (uint128) {
+        if (value > type(uint128).max) revert SafeCastValueDoesNotFit128Bits();
+        return uint128(value);
+    }
+
+    modifier onlyOwner() {
+        if (msg.sender != OWNER) revert NotOwner();
+        _;
+    }
+
+    /// @notice Reverts when the contract is uninitialized
+    modifier whenInitialized() {
+        if (CONTRACT_VERSION_POSITION.getStorageUint256() == 0) {
+            revert Uninitialized();
+        }
+        _;
+    }
+
+    /// @notice Reverts when the caller is not Lido DAO Agent
+    modifier onlyLidoDAOAgent() {
+        if (msg.sender != LIDO_DAO_AGENT_POSITION.getStorageAddress()) {
+            revert LidoDAOAgentExpected(msg.sender);
+        }
+        _;
+    }
+
+    /// @notice Reverts when new withdrawal requests placement resumed
+    modifier whenPaused() {
+        if (REQUESTS_PLACEMENT_RESUMED_POSITION.getStorageBool()) {
+            revert PausedRequestsPlacementExpected();
```

Comment:
The pause trigger should affect:
- new requests placement
- fulfilling the existing ones